spec/models/course_spec.rb

```diff
@@ -142,9 +142,457 @@
     end
   end
 
-  describe '#upload_assignment_list' # TODO
-  describe '#get_required_files' # TODO
+  describe '#upload_assignment_list' do
+    let!(:test_course) do
+      Course.where(name: 'TestCourse-2022-Fall', display_name: 'Test Course')
+            .first_or_create(name: 'TestCourse-2022-Fall', is_hidden: false,
+                             display_name: 'Test Course')
+    end
+    before(:each) do
+      test_course.assignments.destroy_all
+      expect(test_course.assignments.to_a).to eq([])
+    end
+    context 'when file_format = \'csv\'' do
+      context 'when the file contains no assignments' do
+        it 'should not change the state of the database' do
+          # get assignments associated with a course before uploading assignments
+          assignments_before_upload = test_course.assignments.to_ary
+
+          test_course.upload_assignment_list('csv', [].to_csv)
+
+          # reload the course so that we can see if the state of the course has changed in the database
+          test_course.reload
+
+          assignments_after_upload = test_course.assignments.to_ary
+
+          # Expect there to be no stored assignments in the database
+          expect(assignments_after_upload).to eq(assignments_before_upload)
+        end
+      end
+      context 'when the file contains a single assignment' do
+        before(:each) do
+          @assignment_before_call = test_course.assignments.new(short_identifier: 'TEST', is_hidden: false,
```

Comment:
1. Don't use instance attributes, and instead use `let!`.
2. Similar to above comment, you should use a factory here to fill in attributes for you. In this case, you can do `build(:assignment, course: course)` to ensure the assignment is associated with the correct course.

    (Note: `build` will use the factory to create instance, but won't save it to the database, unlike `create`.)